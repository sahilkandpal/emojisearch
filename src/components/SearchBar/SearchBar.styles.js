import styled from "styled-components";
export const StyledSearchBar = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  & {
    .search {
      display: flex;
      border: 1px solid lightgrey;
      border-radius: 50px;
      width: 500px;
      justify-content: center;
      align-items: center;
      padding: 5px 10px;
      background-color: #fff;
      @media screen and (max-width: 576px) {
        width: 80%;
      }
      &-input {
        height: 42px;
        outline: none;
        border: none;
        padding: 0px 10px;
        flex: 1;
        border-radius: 50px;
        font-size: 20px;
        @media screen and (max-width: 576px) {
          width: 90%;
        }
      }
      .cross-icon {
        font-size: 25px;
        border-radius: 50px;
        cursor: pointer;
      }
    }
  }
`;
