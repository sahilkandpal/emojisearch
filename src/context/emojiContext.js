import { createContext, useState, useContext } from "react";
import emojidata from "../json/emoji-en-US.json";
export const emojiContext = createContext({
  emojiData: {},
  filterEmojiData: () => {},
  handleEmojiData: () => {},
  loading: false,
  handleLoading: () => {},
});
const initialState = emojidata;

export const EmojiContextProvider = ({ children }) => {
  const [emojiData, setEmojiData] = useState(initialState);
  const [loading, setLoading] = useState(false);

  const handleLoading = (state) => {
    setLoading(state);
  };

  const handleEmojiData = (val) => {
    setEmojiData(val);
  };

  const filterEmojiData = (query) => {
    if (query.trim() !== "") {
      let newdata = {};
      Object.keys(emojidata).map((val) => {
        const filtered = emojidata[val].filter((value) =>
          value.toLowerCase().includes(query.trim().toLowerCase())
        );
        if (filtered.length > 0) {
          newdata = { ...newdata, [val]: emojidata[val] };
        }
        return null;
      });
      return newdata;
    } else {
      return emojidata;
    }
  };

  return (
    <emojiContext.Provider
      value={{
        emojiData,
        filterEmojiData,
        handleEmojiData,
        loading,
        handleLoading,
      }}
    >
      {children}
    </emojiContext.Provider>
  );
};

export const useEmojiContext = () => {
  const {
    emojiData,
    filterEmojiData,
    handleEmojiData,
    loading,
    handleLoading,
  } = useContext(emojiContext);
  return {
    emojiData,
    filterEmojiData,
    handleEmojiData,
    loading,
    handleLoading,
  };
};
