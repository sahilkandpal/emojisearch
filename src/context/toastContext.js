import { createContext, useState, useContext } from "react";

const initialState = { message: "", isVisible: false };

export const toastContext = createContext({
  toast: initialState,
  handleToast: () => {},
});

export const ToastContextProvider = ({ children }) => {
  const [toast, setToast] = useState(initialState);

  const handleToast = (message) => {
    setToast({ message, isVisible: true });
    setTimeout(() => {
      setToast({ message, isVisible: false });
    }, 1000);
  };

  return (
    <toastContext.Provider value={{ toast, handleToast }}>
      {children}
    </toastContext.Provider>
  );
};

export const useToastContext = () => {
  const { toast, handleToast } = useContext(toastContext);
  return {
    toast,
    handleToast,
  };
};
