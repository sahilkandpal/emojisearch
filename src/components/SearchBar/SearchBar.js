import React, { useState } from "react";
import { StyledSearchBar } from "./SearchBar.styles";
import { useEmojiContext } from "../../context/emojiContext";
import useDebounce from "../../hooks/useDebounce";
const SearchBar = ({ className }) => {
  const [query, setQuery] = useState("");
  const { filterEmojiData, handleEmojiData, handleLoading } = useEmojiContext();
  const debounce = useDebounce();

  const handleDebounce = (value) => {
    handleLoading(true);
    debounce(() => changeEmojiData(value), 500);
  };

  const changeEmojiData = (q) => {
    const newdata = filterEmojiData(q);
    handleEmojiData(newdata);
    handleLoading(false);
  };

  const handleChange = (e) => {
    const value = e.target.value;
    setQuery(value);
    handleDebounce(value);
  };

  const clearValue = () => {
    setQuery("");
    query && handleDebounce("");
  };

  return (
    <StyledSearchBar className={className}>
      <div className="search">
        <input
          data-testid="search-input"
          className="search-input"
          type="text"
          value={query}
          onChange={handleChange}
          autoFocus={true}
          placeholder="Search"
        />
        <div
          data-testid="cross-icon"
          className="cross-icon"
          onClick={clearValue}
        >
          ‚ùå
        </div>
      </div>
    </StyledSearchBar>
  );
};

export default SearchBar;
