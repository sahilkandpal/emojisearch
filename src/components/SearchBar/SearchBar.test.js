import React from "react";
import { shallow } from "enzyme";
import SearchBar from "./SearchBar";
import { useEmojiContext } from "../../context/emojiContext";

jest.mock("../../context/emojiContext");

jest.useFakeTimers("modern");
describe("Given Input", () => {
  let filterEmojiDataMock = jest.fn();
  let handleEmojiDataMock = jest.fn();
  let handleLoadingMock = jest.fn();

  beforeEach(() => {
    useEmojiContext.mockReturnValue({
      filterEmojiData: filterEmojiDataMock,
      handleEmojiData: handleEmojiDataMock,
      handleLoading: handleLoadingMock,
    });
  });

  test("Search bar should render properly", () => {
    const component = shallow(<SearchBar />);
    expect(component).toMatchSnapshot();
  });

  test("should render an input tag", () => {
    const component = shallow(<SearchBar />);
    expect(component.find("[data-testid='search-input']").exists()).toBe(true);
  });

  test("should render an cross icon", () => {
    const component = shallow(<SearchBar />);
    expect(component.find("[data-testid='cross-icon']").exists()).toBe(true);
  });

  test("input should update value on change and set loading true", () => {
    const component = shallow(<SearchBar />);
    let input = component.find("[data-testid='search-input']");
    input.simulate("change", { target: { value: "hot" } });
    input = component.find("[data-testid='search-input']");
    expect(input.props().value).toEqual("hot");
    expect(handleLoadingMock).toHaveBeenCalledWith(true);
    jest.runAllTimers();
  });
  test("input value on change should call some functions after 500ms", () => {
    const component = shallow(<SearchBar />);
    let input = component.find("[data-testid='search-input']");
    input.simulate("change", { target: { value: "face" } });
    input = component.find("[data-testid='search-input']");
    expect(input.props().value).toEqual("face");
    setTimeout(() => {
      expect(filterEmojiDataMock).toHaveBeenCalledWith("face");
      expect(filterEmojiDataMock).toHaveBeenCalledTimes(1);
      expect(handleEmojiDataMock).toHaveBeenCalledTimes(1);
      expect(handleLoadingMock).toHaveBeenNthCalledWith(2, false);
      expect(handleLoadingMock).toHaveBeenCalledTimes(2);
    }, 500);

    jest.runAllTimers();
  });

  test("should clear input value on clear", () => {
    const component = shallow(<SearchBar />);
    let input = component.find("[data-testid='search-input']");
    input.simulate("change", { target: { value: "happy" } });
    let crossIcon = component.find("[data-testid='cross-icon']");
    crossIcon.simulate("click");
    input = component.find("[data-testid='search-input']");
    expect(input.props().value).toEqual("");

    jest.runAllTimers();
  });

  test("input value on clear should call some functions after 500ms", () => {
    const component = shallow(<SearchBar />);
    let input = component.find("[data-testid='search-input']");
    input.simulate("change", { target: { value: "" } });
    input = component.find("[data-testid='search-input']");
    expect(input.props().value).toEqual("");
    setTimeout(() => {
      expect(filterEmojiDataMock).toHaveBeenCalledWith("");
      expect(filterEmojiDataMock).toHaveBeenCalledTimes(1);
      expect(handleEmojiDataMock).toHaveBeenCalledTimes(1);
      expect(handleLoadingMock).toHaveBeenNthCalledWith(2, false);
      expect(handleLoadingMock).toHaveBeenCalledTimes(2);
    }, 500);

    jest.runAllTimers();
  });
});
