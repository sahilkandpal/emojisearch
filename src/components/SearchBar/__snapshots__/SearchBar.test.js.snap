// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given Input Search bar should render properly 1`] = `
<styled.div>
  <div
    className="search"
  >
    <input
      autoFocus={true}
      className="search-input"
      data-testid="search-input"
      onChange={[Function]}
      placeholder="Search"
      type="text"
      value=""
    />
    <div
      className="cross-icon"
      data-testid="cross-icon"
      onClick={[Function]}
    >
      ‚ùå
    </div>
  </div>
</styled.div>
`;
